modules/host/negotiaterevisecontract.go

```diff
@@ -0,0 +1,337 @@
+// overflows and nil values.
+
+import (
+	"errors"
+	"net"
+	"time"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+
+	"github.com/NebulousLabs/bolt"
+)
+
+var (
+	// errBadModificationIndex is returned if the renter requests a change on a
+	// sector root that is not in the file contract.
+	errBadModificationIndex = errors.New("renter has made a modification that points to a nonexistant sector")
+
+	// errBatchCountExceeded is returned if the renter tries to send a batch of
+	// modifications that is larger than the maximum allowed batch.
+	errBatchCountExceeded = errors.New("renter is trying to modify things greater than the max batch count")
+
+	// errIllegalOffsetAndLength is returned if the renter tries perform a
+	// modify operation that uses a troublesome combination of offset and
+	// length.
+	errIllegalOffsetAndLength = errors.New("renter is trying to do a modify with an illegal offset and length")
+
+	// errUnknownModification is returned if the host receives a modification
+	// action from the renter that it does not understand.
+	errUnknownModification = errors.New("renter is attempting an action that the host is not aware of")
+)
+
+// managedRevisionIteration handles one iteration of the revision loop. As a
+// performance optimization, multiple iterations of revisions are allowed to be
+// made over the same connection.
+func (h *Host) managedRevisionIteration(conn net.Conn, so *storageObligation) (bool, error) {
+	// Set the negotiation deadline.
+	conn.SetDeadline(time.Now().Add(modules.NegotiateFileContractRevisionTime))
+
+	// Send the settings to the renter. The host will keep going even if it is
+	// not accepting contracts, because in this case the contract already
+	// exists.
+	h.mu.RLock()
+	settings := h.settings
+	secretKey := h.secretKey
+	blockHeight := h.blockHeight
+	h.mu.RUnlock()
+	err := crypto.WriteSignedObject(conn, settings, secretKey)
+	if err != nil {
+		return false, err
+	}
+
+	// Write the most recent file contract revision transaction.
+	var revisionTxn types.Transaction
+	if len(so.RevisionTransactionSet) > 0 {
+		revisionTxn = so.RevisionTransactionSet[len(so.RevisionTransactionSet)-1]
+	}
+	err = encoding.WriteObject(conn, revisionTxn)
+	if err != nil {
+		return false, err
+	}
+
+	// The renter will either accept or reject the settings + revision
+	// transaction.
+	var acceptStr string
+	err = encoding.ReadObject(conn, &acceptStr, modules.MaxErrorSize)
+	if err != nil {
+		return false, err
+	}
+	if acceptStr != modules.AcceptResponse {
+		return false, errors.New(acceptStr)
+	}
+
+	// The renter is now going to send a batch of modifications followed by and
+	// update file contract revision. Read the number of modifications being
+	// sent by the renter.
+	var modificationCount uint64
+	err = encoding.ReadObject(conn, &modificationCount, 8)
```

Comment:
On the renter side this is simply:
```go
	// send 'delete' action
	err := encoding.WriteObject(he.conn, []modules.RevisionAction{{
		Type:        modules.ActionDelete,
		SectorIndex: uint64(index),
	}})
```
Is there any reason why you couldn't do the same?
```go
var modifications []modules.RevisionAction
err = encoding.ReadObject(conn, &modifications, maxBatchSize)
if err != nil {
    return false, err
} else if len(modifications) > settings.MaxBatchCount {
    return false, errBatchCountExceeded
}

for _, m := range modifications {
    switch m.Type {
    case modules.ActionDelete:
    // ...
    }
}
```